### PR TITLE
feat(ff): add div_rem to BigInteger

### LINKED ISSUE #645: Add multiplication and division methods for `BigInteger`
This will simplify some code, e.g. see [here](https://github.com/arkworks-rs/algebra/pull/644#discussion_r1176874187).

### PR DESCRIPTION
**Pull request by pjdurden**
Closes #645.

`mul` for `BigInteger` was added in #772, but division was still missing, so code like `glv.rs` still has to convert into `num_bigint::BigInt` to divide. This adds a `div_rem` to the trait that returns `(quotient, remainder)` for division by another `BigInteger`.

The implementation reuses the same base-2 long division as the existing `const_modulo!` helper (which only computes the remainder), extended to also accumulate the quotient bit-by-bit. No new arithmetic primitives needed.

Testing: added a randomized test that checks `div_rem` against `num_bigint` as an oracle (quotient and remainder) over 1000 random pairs for every `BigInteger` width, plus edge cases (0/n, n/n, dividend < divisor, division by one). Panics on division by zero, matching `const_modulo!`.

I left the `glv.rs` cleanup out of this PR to keep it focused on the primitive — happy to follow up with that (and a wide-dividend variant) if useful.

### CODE REVIEW

**Comment by pjdurden** [comment]
addressed reviews. `div_rem` is now a default trait method so adding it is non-breaking for existing `BigInteger` implementors; the `isize`-cast and post-move test concerns were already resolved in the latest revision, and all biginteger tests + the doctest pass. cc @Pratyush @davxy

File: ff/src/biginteger/mod.rs [hunk 1]
```diff
@@ -1130,6 +1165,35 @@ pub trait BigInteger:
...
+    /// remainder, i.e. the unique pair `(q, r)` such that
+    /// `self = q * divisor + r` with `r < divisor`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `divisor` is zero.
+    ///
+    /// # Example
+    ///
+    /// ```
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply ff/src/biginteger/mod.rs:1146]: Done. div_rem is now a default trait method built on existing trait primitives (From<u64>, AsMut<[u64]>, mul2, get_bit, sub_with_borrow), so current BigInteger implementors keep compiling without supplying a body.

File: ff/src/biginteger/mod.rs [hunk 2]
```diff
@@ -460,6 +460,41 @@ impl<const N: usize> BigInteger for BigInt<N> {
...
+        // Base-2 long division: the same algorithm as the `const_modulo!`
+        // helper, extended to also accumulate the quotient. We scan the bits of
+        // `self` from the most significant down to the least, shifting each into
+        // a running remainder; whenever the remainder reaches `divisor` we
+        // subtract it and set the corresponding bit of the quotient.
+        let mut quotient = Self::zero();
+        let mut remainder = Self::zero();
+
+        let mut i = self.num_bits() as isize - 1;
+        while i >= 0 {
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply ff/src/biginteger/mod.rs]: The current revision iterates `(0..self.num_bits() as usize).rev()`, so the index is usize throughout and there is no isize cast that can wrap negative.

File: ff/src/biginteger/mod.rs [hunk 3]
```diff
@@ -460,6 +460,41 @@ impl<const N: usize> BigInteger for BigInt<N> {
...
+            remainder.0[0] |= self.get_bit(i as usize) as u64;
+
+            if carry || remainder >= *divisor {
+                let borrow = remainder.sub_with_borrow(divisor);
+                debug_assert_eq!(borrow, carry);
+                quotient.0[0] |= 1;
+            }
+
+            i -= 1;
+        }
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply ff/src/biginteger/mod.rs]: Same here: the loop is an all usize countdown via `(0..self.num_bits() as usize).rev()`, so no isize cast remains.

File: ff/src/biginteger/tests.rs [hunk 4]
```diff
@@ -227,6 +227,46 @@ fn biginteger_conversion_test<B: BigInteger>() {
...
+        }
+
+        let (q, r) = a.div_rem(&b);
+
+        let a_big: BigUint = a.into();
+        let b_big: BigUint = b.into();
+        assert_eq!(Into::<BigUint>::into(q), &a_big / &b_big);
+        assert_eq!(Into::<BigUint>::into(r), &a_big % &b_big);
+        // The remainder is always strictly smaller than the divisor.
+        assert!(r < b);
```
[the remark this answers is not in the document]
Comment (pjdurden) [reply ff/src/biginteger/tests.rs]: The test now asserts `r < b` before the BigUint conversions, so r is never read after the Into move.

**Review by weikengchen — CHANGES_REQUESTED** [review]
Thanks for putting this together. I’m concerned that this PR is unlikely to be merged in its current shape.

The main issue is that this adds a runtime `BigInteger` division primitive implemented as base-2 long division. While the implementation looks correct, I do not expect this to outperform `num_bigint`’s division in realistic use. `num_bigint::BigInt` is not cheap, but it is at least doing limb-oriented arithmetic; bit-by-bit long division is expected to be worse.

I also think the motivating GLV cleanup probably still cannot use this method directly, because the GLV case divides a wide/intermediate value rather than a same-width `BigInteger` dividend. So this PR adds a general primitive, but likely does not remove the `num_bigint` dependency from the high-value path.

`const_modulo!` is a different case: it is compile-time machinery where simplicity and const-evaluability matter more, so reusing that algorithm for a runtime division API has a different performance story.

My suggestion would be to retarget this toward the GLV use case directly: replace the current `num_bigint` division there with a specialized, efficient routine for the needed wide/signed division shape. That seems much more likely to justify the complexity and give arkworks a real win.

(drafted by GPT 5.5)

**Comment by weikengchen** [comment]
One concrete method to consider for the GLV path is to avoid generic division entirely and target the exact shape we need there:

```text
beta = round(k * coeff / r)
```

Here `r` is the scalar-field modulus and `coeff` is fixed per curve, so we can precompute a fixed-point/scaled reciprocal-style constant, for example:

```text
g = round(2^S * coeff / r)
```

Then at runtime compute roughly:

```text
beta = round(k * g / 2^S)
```

with enough precision in `S` and, if necessary, a small correction step. The sign of `coeff` can be handled separately.

This turns the expensive division into wide multiplication plus a shift, and it is much closer to the actual GLV scalar-decomposition need than a general same-width `BigInteger::div_rem`. It also has a better chance of beating the current `num_bigint` route, whereas base-2 long division is likely to be slower.

**Comment by pjdurden** [comment]
Thanks, that is fair and I would rather not ship something that does not earn its place.

You are right on both counts. Base-2 long division is bit-at-a-time, so against num_bigint's limb-oriented division it loses at the widths that matter, and I am not going to pretend a benchmark would say otherwise. And the GLV path divides k * coeff, which is wider than a single BigInteger, so a same-width div_rem would not remove the num_bigint conversion there anyway. So this PR adds a primitive that is both slower and does not unblock the high-value path. Agreed it should not merge as-is.

I will go with option 1: close this PR and open a focused one that does just the GLV reciprocal decomposition, no public BigInteger::div_rem.

The plan there: r and coeff are fixed per curve, so precompute a scaled reciprocal at compile time

    g = round(2^S * coeff / r)

and at runtime compute beta = (k * g) >> S, then a small correction step (a couple of conditional subtractions of r) to land on the exact rounded value. Pick S so the truncation error over the full range of k stays below one, which keeps the correction bounded and cheap. That replaces the num_bigint division on the actual hot path with a wide multiply and a shift.

Will close this once the GLV PR is up so the discussion has somewhere to land. Thanks for the steer.

**Comment by pjdurden** [comment]
Closing as discussed. The GLV path divides `k * n22` by `r`, and that product is wider than a single `BigInt<N>`, so this same-width primitive would not remove the `num_bigint` conversion in `scalar_decomposition` anyway.

I have not got the scaled-reciprocal version benchmarked against the current path yet. Will open it as a separate PR if it wins.

**Comment by weikengchen** [comment]
Thanks a lot. I am tight on my ChatGPT Codex usage limit, and I have been curious if ChatGPT can code a good implementation with a simple prompt.
